### ISSUE TITLE
`restrict`: `embed`'s adjoint — slicing back down to contained graded spaces

### ISSUE THREAD

**Issue by Ryo-wtnb11** [Ryo-wtnb11/TeNeT-py#90]
**Goal**

Ship the direction #83 deliberately left out, and settle the question it deferred.

`tenet.embed(t, legs)` (#83) grows a graded bond: same data in the leading degeneracy slots, zeros everywhere else. Its inverse — keep the leading `m_a` slots of each sector, drop sectors that left the target space — is mechanically five lines, and #83's own round-trip criterion already carries them, by hand, in `tests/ops/test_embed.py::project`:

```python
def project(embedded, source):
    """The restriction, by hand: the leading degeneracy slots of each source key.

    Five lines, and the reference implementation for a future ``restrict`` —
    which is not shipped because its refusal semantics (what happens to non-zero
    data in a dropped slot) have no caller to settle them.
    """
```

There is now a caller, and it settles them. The VMC/PEPS shape is: grow a bond with `embed`, run an update that writes into the new slots, then project back down to the working bond space. In that loop the discarded slots are *not* zero — they are physics — and a function that silently drops them is the one shape guaranteed to produce a wrong answer nobody notices. So `restrict` refuses above a tolerance, on exactly the pattern `from_dense` (#82) already established: relative default `atol`, `atol=math.inf` as the documented traceable escape.

Prototyped in ten lines over basic slicing and `_check_containment` with its arguments swapped. Measured on the 3-leg SU(2) fixture pair from #83 (`{j=0: 2, j=1: 1}` ↔ `{j=0: 3, j=1: 2, j=2: 1}`): `restrict(embed(t, L), t.legs)` reproduces `t.blocks` at `0.0` with residual exactly `0.0`; `embed(restrict(u, S), u.legs)` is idempotent at `0.0`; Pythagoras `‖u‖² = ‖restrict(u)‖² + residual²` holds at `0.0` (residual `11.5830` against `‖u‖ = 12.5849`); and adjointness `⟨embed(a), b⟩ = ⟨a, restrict(b)⟩` in the `qdim`-weighted inner product agrees to `4.4e-16`.

**Design / API sketch**

```python
# src/tenet/ops/embed.py — the sibling lives next to embed, in the same module.
# Exported as `tenet.restrict`, plus a method.

def restrict(
    t: SymmetricTensor, legs: Sequence[Leg], *, atol: float | None = None
) -> SymmetricTensor:
    """``t`` re-expressed on smaller legs: the leading degeneracy slots, nothing else.

    The exact mirror of ``embed``: ``legs[i]`` must match ``t.legs[i]`` in provider,
    ``side`` and ``dual``, and its space must be **contained in** ``t.legs[i].space``
    — every sector ``a`` of the target appears in the source with
    ``degeneracy(a) >= target.degeneracy(a)``. Sectors absent from ``legs`` are
    dropped entirely. ``name`` is taken from ``legs`` and never compared.

    Placement is the same **prefix in the degeneracy index** ``embed`` uses, so the
    two agree about which slots are the "old" ones, and so does a subsequent
    ``svd(..., bond=B)`` (#77).

    Data outside the kept slots is **refused**, not discarded: the residual
    ``sqrt(‖t‖² - ‖restrict(t)‖²)`` is compared against ``atol``, which defaults to
    ``sqrt(eps(dtype)) * ‖t‖`` — the relative spelling ``from_dense`` uses (#82),
    for the same units reason. ``atol=math.inf`` projects without checking, and
    that is the form that goes inside ``jit``.
    """
```

The whole body:

```python
def restrict(t, legs, *, atol=None):
    new = TensorStructure(tuple(legs))
    _check_containment(new, t.structure, op="restrict")   # embed's checker, reversed
    index = {k: i for i, k in enumerate(t.structure.block_order)}
    blocks = tuple(
        t.blocks[index[key]][tuple(slice(0, m) for m in new.block_shape(key))]
        for key in new.block_order
    )
    out = SymmetricTensor(new, blocks)
    if atol != math.inf:
        _refuse_discarded(t, out, atol)     # concrete-value question; see below
    return out
```

```python
# tensor.py — one-line delegation, function-local import, as every other op already is
def restrict(self, legs: Sequence[Leg], *, atol: float | None = None) -> "SymmetricTensor": ...
```

**Key decisions & rationale**

- **The deferred question is answered "refuse", and `from_dense` is the precedent, not an analogy.** #83 listed the three candidates and declined to pick: silent discard, discard plus a returned residual, or refusal above a tolerance. With the VMC/PEPS caller in hand, silent discard is out immediately — the *whole point* of the caller's loop is that the dropped slots carry the growth the update just produced, so silence there is silence about the thing being measured. A returned residual is out because it changes the return type of the common path (`u, ()` unpacking at every call site) to serve the rare one, and because the residual is one `tenet.norm` subtraction the caller can compute — and *is* exactly what the refusal message already reports. That leaves refusal, and #82 already shipped its exact shape: relative `atol`, a message naming residual/tolerance/worst offender, and `math.inf` as the escape. Copying that shape character for character is cheaper than inventing a second convention and means a caller who has read one refusal has read both.
- **The default `atol` is relative, and for the same units reason `from_dense` gives.** `sqrt(eps(dtype)) * ‖t‖`. An absolute tolerance is meaningless for an unnormalized tensor: the same wavefunction scaled by `1e6` would be refused or accepted depending on nothing. `from_dense`'s default is literally `math.sqrt(eps) * math.sqrt(total)`; `restrict` uses the same expression against `‖t‖`, and a criterion below asserts the scale invariance directly.
- **The residual is exact by Pythagoras, and costs two `norm` calls — no new machinery, and no per-block accounting on the fast path.** `ops.basic.norm` is `sqrt(Σ_τ qdim(c_τ) ‖A_τ‖²)`. Restriction keeps a sub-block of each retained key and drops whole keys, so the kept and discarded parts occupy disjoint slots under the *same* `qdim` weight: `‖t‖² = ‖restrict(t)‖² + ‖discarded‖²`, exactly. Measured at `0.0` (not `allclose`) in the prototype. The refusal message additionally names the worst offending `FusionBlockKey`, which needs a per-key accumulation — but that loop runs only on the check path, which is untraceable anyway, so nothing on the `atol=math.inf` path pays for it.
- **`restrict` is traceable exactly when the check is off, and it is `embed`'s adjoint in the AD sense too — for free.** The projection is `block[tuple(slice(0, m) for m in shape)]`, pure basic slicing, which every backend traces and differentiates; its own VJP is a zero-pad, i.e. `embed`. So `jax.grad` through `restrict` produces cotangents with exactly the *source* block shapes, zeros in the dropped slots — a criterion, and the cleanest possible statement that the two operations are adjoint. The tolerance comparison is a concrete-value question and raises `ConcretizationTypeError` under a trace, in JAX's own voice, exactly as `from_dense`'s default `atol` and `SymmetricTensor.__eq__` already do. Two behaviours, one screen, one test.
- **`_check_containment` is reused with its arguments swapped, plus an `op` name — not copied, and not generalized.** #83's checker already walks leg count, provider, `side`, `dual` and per-sector degeneracies, and containment of A in B is exactly what `restrict` needs with A and B exchanged. The only edit is threading an `op: str` through the messages so `"embed: axis 2 sector …"` can read `"restrict: …"`, and rewording the two trailing clauses that currently say "embed never truncates" / "shrinking is a different operation". One parameter, no new function, no `Containment` helper class. `ponytail:` shared checker; split it only if the two refusals ever need to say genuinely different things.
- **`restrict` is *not* a truncation, and the distinction is the same one #64 and #77 drew.** The target space is chosen by the caller from metadata, so `restrict` is shape-static and traceable, and it sits with `embed` and `svd(..., bond=B)` on the traceable side of #64's `StructureChangingError`. A restriction whose target is *decided from the block values* — "drop whatever falls below `1e-8`" — is `svd_truncated`'s job and stays non-traceable there. A caller who wants to shrink a bond *because the data says so* should not reach for `restrict`; the docstring says so and points at `svd_truncated`.
- **Placement stays the leading degeneracy slots, and it is not a prefix of the dense index — same trap as #83.** The within-slab dense layout is `alpha * d_a + m`, so for SU(2) the kept data is strided in the dense array and every later sector's slab offset moves. Any dense-side oracle must be written through the `(alpha, m)` decomposition; `tests/ops/test_embed.py::index_maps` already exists for exactly this and is reused rather than rewritten.
- **No `restrict` for `side`, `dual` or provider, and no "closest contained space" repair.** All three are the refusals `embed` already raises, for the same reasons: `side`/`dual` are categorical (invariant 2) and moving between them is `repartition`; a provider change is a cast, filed alongside this one. And a function that sometimes restricts and sometimes silently picks a different target space is the failure mode this whole pair of issues is shaped to avoid.
- **Round-trip laws, and which one is exact.** `restrict(embed(t, L), t.legs) == t` is **exact** (`0.0`, and the default check passes with residual `0.0`) because `embed` writes zeros into precisely the slots `restrict` drops. The other order is a *projection*, not an identity: `P = embed ∘ restrict` satisfies `P² = P` (measured at `0.0`) and `‖u‖² = ‖P u‖² + residual²`, and it is orthogonal — which is what `⟨embed(a), b⟩ = ⟨a, restrict(b)⟩` (agreeing to `4.4e-16` in the prototype) actually says. All three are criteria, because "adjoint" in this issue's title should be an executable claim rather than a word.

**References**

- README: "Fusion spaces" / "`GradedSpace`", "Structure-changing differentiation" (the traceable side of the line, for the same reason `embed` is), "Public axis order", "Linear algebra"; invariants **2**, **5**, **8**, **9**, **11**.
- Code: `src/tenet/ops/embed.py` (#83) — `embed`, `_check_containment` (reused with swapped arguments), the prefix convention, and the two `ponytail:` notes this issue discharges; `tests/ops/test_embed.py` (#83) — `project()`, the five-line reference implementation this promotes, and `index_maps()`, the `(alpha, m)` dense oracle reused unchanged; `src/tenet/ops/dense.py` (#82) — `from_dense`'s `atol` default `math.sqrt(eps) * math.sqrt(total)`, its `math.inf` escape, and `_check_symmetric`'s message shape (residual, tolerance, worst offender), all copied here; `src/tenet/ops/basic.py` (#20) — `norm`'s `qdim` weight, which makes the Pythagoras identity exact; `src/tenet/space.py` (#5) — `GradedSpace`, `degeneracy` returning `0` for an absent sector, `sector_offset`/`dim` for the dense oracle, and `ProductSpace.matches`'s "`name` is user bookkeeping" stance; `src/tenet/leg.py` (#6); `src/tenet/structure.py` (#8) — `block_order`, `block_shape`, `index_of`, and `_block_order`'s sector-assignment enumeration, which is what makes the target's keys a *subset* of the source's; `src/tenet/ops/linalg.py` (#64, #77) — `svd_truncated`, the data-decided truncation this is emphatically not, and `StructureChangingError`; `src/tenet/pytree.py` (#41).
- **TensorKit.jl's `⊖` / `ominus`** — https://github.com/QuantumKitHub/TensorKit.jl , `src/spaces/vectorspaces.jl` — the *space*-level complement satisfying `V == W ⊕ (V ⊖ W)`, documented at https://jutho.github.io/TensorKit.jl/stable/man/spaces/ . TensorKit has no tensor-level restriction; the nearest thing is `absorb!` in `src/tensors/linalg.jl`, which copies a tensor into a differently-sized one overwriting only the shared region — i.e. silent discard, the option this issue rejects.
- **froSTspin's `check_norm`** — https://github.com/ogauthe/frostspin , `src/frostspin/symmetric_tensor/tools.py` — the relative `tol * max(n1, n2)` comparison that #82 adopted the scaling from, and which merely `warnings.warn`s where we raise.

**Out of scope**

- Deciding the target space from the block values ("drop sectors whose weight is below `x`"). That is `svd_truncated`'s territory (#64), it is non-traceable by design, and `restrict`'s docstring points at it rather than growing a mode.
- Any placement other than the leading degeneracy slots — a general index selection, a `slots=` argument, keeping the *largest* rather than the *first* entries. The prefix convention is shared with `embed` and `svd(..., bond=)` and this issue does not open it.
- Returning the residual, a `(tensor, residual)` tuple, or a `residual=` out-parameter. The refusal message reports it; a caller who wants the number computes `sqrt(norm(t)**2 - norm(restrict(t, legs, atol=inf))**2)`, which is the same line the implementation runs.
- Making the discarded-mass check traceable. It is a concrete-value question; `atol=math.inf` is the documented escape, exactly as in #82.
- Changing `side`, `dual` or provider while restricting. Refusals, not conversions.
- A plan object or a cache. Same argument as #83: a dict lookup and a slice per block.
- A `TensorNetwork`-level or `quimb`-level bond-shrinking helper.
- PyTorch; custom VJPs (a slice is linear and every backend differentiates it); autoray registration under the `"tenet"` backend — that list stays closed (invariant 11).

**Acceptance criteria**

- `tenet.restrict(t, legs, *, atol=None)` exists in `src/tenet/ops/embed.py`, is exported from `tenet` and `tenet.ops`, and `SymmetricTensor.restrict(legs, atol=None)` is a one-line delegation with a function-local import.
- Target keys are a subset of source keys: `set(restrict(t, legs).structure.block_order) <= set(t.structure.block_order)`, asserted directly for every fixture pair — the fact the loop rests on.
- **Exact round-trip with `embed`:** `restrict(embed(t, L), t.legs)` equals `t` block-for-block at **`0.0`** (not `allclose`) and its residual is exactly `0.0` so the default check passes, for Trivial, U(1), SU(2), fZ2 and a product symmetry, including an `L` that adds an entirely new sector and one that only grows existing degeneracies. Measured `0.0` / `0.0` in the prototype.
- **`P = embed ∘ restrict` is an idempotent orthogonal projection:** `P(P(u))` equals `P(u)` at `0.0`, and `⟨embed(a), b⟩ == ⟨a, restrict(b, atol=inf)⟩` in the `qdim`-weighted inner product to `1e-14`, for U(1) and SU(2). Measured `0.0` and `4.4e-16`.
- **Pythagoras is exact:** `norm(t)**2 == norm(restrict(t, legs, atol=inf))**2 + residual**2` to `1e-14`, for all four providers plus a product symmetry, including SU(2) where the `qdim` weight makes it a non-trivial statement. Measured `0.0` on SU(2) with residual `11.5830` against `‖u‖ = 12.5849`.
- **Non-zero discarded data is refused:** a tensor with non-zero content in a slot being dropped, and one with a whole sector being dropped, each raise `ValueError` naming the residual, the tolerance and the worst offending `FusionBlockKey` with its discarded mass. A discarded mass below `atol` is accepted silently.
- The default `atol` is relative and asserted to scale: the same relative discarded fraction is refused identically on `t` and on `1e6 * t`.
- **Structural refusals**, raised before any block is read and identically inside and outside `jit`, each a `ValueError` naming the axis and both degeneracies where relevant, and each asserted on message content: a target sector with *larger* degeneracy than the source's; a target sector absent from the source; a `side` mismatch; a `dual` mismatch; a provider mismatch; a leg-count mismatch. The messages say `restrict:`, not `embed:`, and `tests/ops/test_embed.py`'s existing `embed` refusal messages are unchanged.
- **Dense oracle, written through `(alpha, m)`:** `restrict(t, legs, atol=inf).to_dense()` equals `t.to_dense()` indexed by `tests/ops/test_embed.py::index_maps(legs, t.legs)` to `0.0`, required for SU(2) specifically where `d_a > 1` makes the kept data strided rather than a dense prefix.
- **Traceability, both sides, in one screen:** `jax.jit(lambda t: tenet.norm(tenet.restrict(t, legs, atol=math.inf)))` compiles and traces once per source structure (changing `legs` retraces, changing block values does not); the same call with the default `atol` raises `ConcretizationTypeError` — JAX's own error, not ours. The test sits next to #64's `StructureChangingError` test and #82's `from_dense` tolerance test, with a comment naming the #77 static-metadata argument.
- **`restrict` is `embed`'s adjoint under AD:** `jax.grad` of a scalar built from `restrict(t, legs, atol=math.inf)` matches central finite differences to `1e-6` in float64 for U(1) and SU(2), and the cotangent has exactly the *source* block shapes with exact zeros in the dropped slots.
- Complex dtypes: a `complex128` tensor restricts to `complex128` with exact block equality on the kept slots, and the residual of a complex discard is computed from `|·|²`.
- Backend-generic: the same criteria pass with NumPy and JAX blocks; `restrict` on a JAX-backed tensor returns JAX blocks (`ar.infer_backend`); no `import jax` in `ops/embed.py` and `test_only_pytree_module_imports_jax` passes unchanged, still asserting equality against `{"pytree.py", "ad.py"}`.
- `tests/ops/test_embed.py::project`'s comment is updated to point at `restrict` instead of describing it as unshipped, and `embed`'s round-trip criterion is re-expressed through `restrict` while keeping the hand-written projection as the independent oracle (the test must not verify `restrict` against `restrict`).
- The `ponytail:` note in `ops/embed.py` recording "no `restrict`" is replaced by one recording the shared `_check_containment` and its split trigger. No plan object and no cache is introduced.
- `array/dispatch.py` is unchanged; its closed-list assertion in `tests/array/test_dispatch.py` passes untouched.
- `ruff check`, `ruff format --check` and the full `pytest` suite pass.

**Depends on**

**#83** (`embed`, `_check_containment`, the prefix convention, and the `tests/ops/test_embed.py::project` reference implementation this promotes), **#82** (`from_dense`'s relative-`atol` / `math.inf` refusal pattern, copied character for character, and its `to_dense` for the oracle), #5 (`GradedSpace`, `degeneracy` returning `0` for an absent sector), #6 (`Leg`, the independent `side`/`dual`), #8 (`TensorStructure`, `block_order`, `block_shape`, `index_of`), #9 (`to_dense` and its `alpha * d_a + m` layout), #20 (`norm`'s `qdim` weight — what makes the Pythagoras residual exact), #41 (`tenet.pytree` — the jit/grad criteria run through it), #64 and #77 (the "static metadata vs. data-decided structure" boundary this lands on the traceable side of, and the `svd_truncated` this deliberately is not).

Independent of the direct-sum and symmetry-cast issues filed alongside this one.
